statsmodels/sandbox/examples/gam_gsoc2015/gam_example.py

```diff
@@ -0,0 +1,360 @@
+                res_aux = self._fit_zeros(keep_index, **kwds)
+                return res_aux
+            else:
+                return res
+
+
+class Penalty(object):
+    """
+    A class for representing a scalar-value penalty.
+    Parameters
+    wts : array-like
+        A vector of weights that determines the weight of the penalty
+        for each parameter.
+    Notes
+    -----
+    The class has a member called `alpha` that scales the weights.
+    """
+
+    def __init__(self, wts):
+        self.wts = wts
+        self.alpha = 1.
+
+    def func(self, params):
+        """
+        A penalty function on a vector of parameters.
+        Parameters
+        ----------
+        params : array-like
+            A vector of parameters.
+        Returns
+        -------
+        A scalar penaty value; greater values imply greater
+        penalization.
+        """
+        raise NotImplementedError
+
+    def grad(self, params):
+        """
+        The gradient of a penalty function.
+        Parameters
+        ----------
+        params : array-like
+            A vector of parameters
+        Returns
+        -------
+        The gradient of the penalty with respect to each element in
+        `params`.
+        """
+        raise NotImplementedError
+
+
+class GamPenalty(Penalty):
+    
+    def __init__(self, wts=1, alpha=1, cov_der2=None, der2=None):
+        
+        self.wts = wts #should we keep wts????
+        self.alpha = alpha
+        self.cov_der2 = cov_der2
+        self.der2 = der2
+    
+    def func(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        '''
+        
+        # The second derivative of the estimated regression function
+        f = np.dot(self.der2, params)
+
+        return self.alpha * np.sum(f**2)
+
+    def grad(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        3) cov_der2 is obtained as np.dot(der2.T, der2)
+        '''
+                
+        return 2 * self.alpha * np.dot(self.cov_der2, params) 
+    
+    def deriv2(self, params):
+
+        return 2 * self.alpha * self.cov_der2
+
+
+class L2(Penalty):
+    """
+    The L2 (ridge) penalty.
+    """
+
+    def __init__(self, wts=None):
+        if wts is None:
+            self.wts = 1.
+        else:
+            self.wts = wts
+        self.alpha = 1.
+
+    def func(self, params):
+        return np.sum(self.wts * self.alpha * params**2)
+
+    def grad(self, params):
+        return 2 * self.wts * self.alpha * params
+
+
+
+class LogitGam(PenalizedMixin, Logit):
+    pass
+
+
+
+class LinearGam(PenalizedMixin, RLM):
+    pass
+
+n = 200
+data = pd.DataFrame()
+x = np.linspace(-10, 10, n)
+poly = x*x 
+y = 1/(1 + np.exp(-poly))
+y[y>0.5] = 1
+y[y<=0.5] = 0 
```

Comment:
this example doesn't really work.
I get only y==1 observations
???

I don't understand what logit below is doing with this y.